### ISSUE TITLE
Experiment: can one track predict its own placement? (PyTorch, single-track features)

### ISSUE THREAD

**Issue by pokle** [pokle/glidecomp#692]
**Status:** plan, 2026-09-08. Experiment, not product code. Nothing here is
imported by the app, the engine's scoring closure, or CI.

## The question

GAP is relative by construction: your points depend on what everybody else
did that day. This experiment asks whether a network that sees **one track at
a time** — never the field — can predict where that track will place.

Two framings, and the second is the real one:

- **Retrospective.** The whole flight is visible. This measures how much of
  placement is explained by the flight itself, once the day and the task are
  known. Expect it to be strong, and largely carried by distance flown; its
  job is to be the reference number the interesting question is read against.
- **Prospective.** Everything the pilot's track carries **up to the moment
  they cross the start cylinder**, plus the task and the day's forecast.
  Nothing after the start. This is the one worth running: can the climb up,
  the wait, and the way a pilot enters the start tell you how the race ends?

## Non-negotiables

- **No pilot identity as a feature.** No CIVL/FAI ranking, no name, no
  federation id, no historical form. Pilot identity is used for two things
  only: grouping, and one clearly-labelled *analysis-only* reference number
  (§ Baselines).
- **No field-derived features.** Every feature is computable from one IGC
  file, the task, and a weather forecast. That rules out the engine's
  `gaggle.*`, `climb.shared_percentile`, `glide.ld_vs_field`,
  `race.time_behind`, and the whole `day.*` family, all of which read the
  other pilots. The extractor therefore builds its own feature set from the
  single-track detectors rather than filtering `FieldContext`.
- **The field appears only in the label.** Placement is by definition
  relative; that is the thing being predicted, not an input.

## Data

Every non-synthetic competition in both repos. `kosci-loop` and `big-chip`
are generated by `web/scripts/generate-*.ts` and are excluded;
`web/samples/comps/corryong-cup-2021-open-t1` is a parity fixture duplicating
an archive task folder and is de-duplicated against it.

| Repo | Comps | Tasks | IGC files |
|---|---|---|---|
| `glidecomp-archive` | 25 | 196 | 5 310 |
| `glidecomp` (`corryong-cup-2026`, `unungra-cup-2020`) | 2 | 10 | 214 |

2017–2026, Australian comps: Bright Open and Unungra (PG), Corryong Cup,
Dalby Big Air and Forbes Flatlands (HG). Roughly 5 500 tracks over 206
task-classes, ~27 per task. Wing comes from `airscore-result-raw.json`'s
`task.comp_class`, with the registry in `download-airscore-comp.ts` as the
fallback.

Tracks a HARD `assessTrackQuality` check withholds are dropped from the
dataset rather than seated at 0 — a broken logger is not a bad flight, and
the count of dropped tracks is reported. This mirrors
`verify-airscore-parity.ts`, which is also where the gap-params merge below
comes from.

## Label

Per task-class, score the whole field with the engine — `scoreTask` under
`resolveCompGapParams(category, {...comp.gap_params, ...task.gap_params})`,
exactly the seed → scorer path — and turn each pilot's total into:

- `percentile` ∈ [0, 1], the primary regression target (1 = task winner);
- `rank`, for the within-task ranking loss;
- `top_decile` / `top_quartile`, the "placed well" classification heads.

Published AirScore totals (`airscore-result-raw.json`, `row[16]`) are carried
alongside as a parity column, so a task where the engine and AirScore
disagree can be spotted rather than silently trained on.

### Track (single-track only)

From `parseIGC` → `detectFlight` / `detectCircles` /
`resolveTurnpointSequence` / `assessTrackQuality`, all of which take one
track. Grouped so the ablation can switch whole blocks off:

- **Climb** — climb-rate distribution (mean, median, p75, p90, best sustained
  10 min), thermal count, mean gain per thermal, time-to-establish, circling
  efficiency and radius, left/right consistency, height band worked
  (min/max/median working height, fraction of time in the top third).
- **Glide** — glide ratio distribution, mean and best glide speed, dolphin
  fraction, time gliding vs circling vs searching, longest glide.
- **Route** — deviation from the optimised course line, extra distance flown
  over optimised, turnpoint entry efficiency, leg speeds as a fraction of
  the leg's optimised distance over elapsed time.
- **Day-from-the-track** — per-circle wind estimate (speed, direction,
  and its component along each leg), drift, altitude ceiling reached, how
  the climb rate trends over the flight.
- **Start** — height and time at start-gate crossing relative to the gate,
  climb rate in the last 10 minutes before it, height relative to the
  pilot's own ceiling so far, number of gates elapsed, distance from the
  start cylinder's optimal exit point.
- **Quality** — the three SOFT `assessTrackQuality` findings and the fix
  interval, as covariates rather than filters.

The prospective variant recomputes every one of these over the truncated
track (takeoff → start crossing) and drops the blocks that are then empty.
Truncation happens **before** the detectors run, not after, so no post-start
fix can leak through a detector's window.

### Task

Optimised distance, leg count and leg-length distribution, turn angles,
cylinder radii, goal type, start gate grid (interval and count), task
deadline, elapsed window, launch elevation and the route's elevation
profile, nominal distance and time from the resolved gap params, wing.
All known before anybody launches.

### Weather

Via the engine's own `web/engine/src/weather/` providers with a disk-cached
`fetch` injected, keyed on `weatherQueryKey` — so the experiment and the app
derive the identical query for a task and the cache is warm for both.

Both datasets:

- **ERA5** (`archive-api`) for all 27 comps — uniform 2017–2026, surface
  wind/gust/temperature/dewpoint, cloud, radiation, boundary-layer height.
  Confirmed reachable right now.
- **Archived forecast** (`historical-forecast-api`) for 2022+, which adds
  pressure-level winds (950–700 hPa) and CAPE. This host answers
  `Daily API request limit exceeded` from the cloud container, whose proxy IP
  is shared, so the fetch runs from a local machine instead. The fetcher is
  resumable and writes straight into the archive store, so it can be stopped
  and restarted, and what it has already fetched is never refetched.

Features are taken over the task's own window: wind speed and direction at
the task's mid-hour and its trend, the wind component along each leg and
along the whole course line, gust ratio, boundary-layer depth at the task
mid-hour and its daily peak, LCL, cloud cover, radiation, CAPE where the
dataset has it. `WeatherSource.variables` says what actually arrived, and a
missing variable is an explicit missing-indicator column, never a zero.

The forecast-only block is an ablation arm: the headline model reports with
and without it, so "did the forecast earn its keep" is answerable rather
than assumed.

## Model

A per-track scoring function `f(track, task, weather) → scalar`. The network
never sees more than one track. The **loss** is what knows about the field:

- **Listwise (headline)** — a Plackett-Luce / ListNet softmax over the tracks
  of one task-class, so the model is trained to order a field it cannot see.
  This is the honest formulation of the question.
- **Pointwise (comparison)** — plain regression on `percentile`, plus a
  `top_decile` classification head, for calibration and for a probability
  that reads as "how likely is this track to place well".

Architecture, in the order it gets built:

1. **Feature MLP.** ~100 standardised engineered features → 2–3 hidden
   layers, wing as an embedding, dropout, missing-value indicators.
2. **Sequence branch**, added once (1) shows signal. The track resampled to
   256 uniform steps × ~8 channels — along-course progress, height above the
   route's terrain reference, climb rate, ground speed, bearing minus course
   bearing, distance to next turnpoint, circling indicator, time of day — into
   a small 1D CNN, its embedding concatenated with the engineered features.
   Reported as a delta against (1): does the raw track carry anything the
   features flatten?

CPU-only. A model at this size trains in minutes, so the run is
cross-validation-bound rather than compute-bound.

## Evaluation

**Grouped by competition, always.** 5-fold cross-validation holding out whole
comps, stratified so each fold sees both wings. A task never straddles the
split, and neither does a comp — otherwise the model learns the day rather
than the flying, and the number means nothing.

Reported per fold and pooled:

- **Spearman ρ between predicted and actual placement, computed within each
  task and then averaged.** This is the headline: it is exactly "did it get
  the order right", and it is immune to the model being globally miscalibrated.
- Precision@3 and @10 for "who won this task".
- AUC for top-quartile.
- MAE on percentile, and a calibration curve for the classification head.

### Baselines the model must beat

1. **Constant** — predicts the mean; ρ = 0 by construction.
2. **Distance flown alone** — one feature, ordinary least squares. If the
   retrospective network does not clearly beat this, the retrospective
   experiment's answer is "GAP is mostly distance", which is worth saying
   plainly.
3. **Made goal + time** — the two-feature scorer's-eye view.
4. **Pilot's historical mean placement** — *analysis only, never a model
   input*, computed leave-one-comp-out. Not a competitor: it is the answer to
   "how much of placement is just who the pilot is", and it is the number
   that tells us whether a track-only model is being asked to do something
   possible. If pilot identity explains most of it, that is the finding.

## Where the data lives

Both the weather answers and the extracted features are **committed to
`pokle/glidecomp-archive`**, beside the tracks they derive from. They are
expensive to produce (the forecast host is rate-limited; the extractor reads
5 500 IGC files) and cheap to store, and every later experiment — this one's
reruns included — then starts from a warm, identical corpus with no network
at all.

```
glidecomp-archive/
  weather/
    README.md
    <task-dir>/era5.json              one file per provider per task,
    <task-dir>/archived-forecast.json   e.g. bright-open-2024-open-t1/
  features/
    README.md
    manifest.json                     extractor version, engine commit,
                                      feature list, per-comp counts
    <task-dir>.jsonl                  one row per track
```

Keyed by task folder, one-to-one with `comps/`, so a file is findable by
eye. Each weather file records the `weatherQueryKey` it answers, so a
consumer can check it against the query the engine would derive for that
task today — a mismatch (a task's turnpoints or times changed) means refetch
rather than silently serve the wrong sky. Each feature row carries the
extractor version, so a feature-set change is visible rather than mixed in.

Rough size: ~6 MB of weather, ~10 MB of features. Both are derived data, but
derived from a source that is rate-limited and, for the forecast archive,
genuinely perishable.

**The features file carries no pilot identity** beyond what is needed to join
back to a track: the IGC filename. No name, no federation id, no ranking.

## Deliverables

In `pokle/glidecomp`:

```
experiments/track-placement/
  PLAN.md                  this plan
  README.md                how to run it, end to end
  extract-features.ts      bun; comps -> features.jsonl (single-track only)
  fetch-weather.ts         bun; engine providers, writes the archive store
  build-dataset.py         archive store -> tensors, folds, standardisation
  train.py                 PyTorch: MLP, +sequence branch, listwise/pointwise
  baselines.py             the four above
  report.py                metrics, ablation table, charts
docs/2026-09-08-track-placement-nn.md   the findings, written up
```

In `pokle/glidecomp-archive`: the `weather/` and `features/` trees above.

Both scripts take the usual `GLIDECOMP_COMPS_DIR` override, and a matching
`GLIDECOMP_ARCHIVE_DIR` for where to write, so a checkout pair anywhere on
disk works. `fetch-weather.ts` is resumable and never refetches what the
store already holds — which is what makes running it from a local machine,
past the shared proxy's rate limit, a matter of leaving it going.

Python is `uv run` with PEP 723 inline dependencies, per the project rule —
torch from the CPU index, no `requirements.txt`, nothing installed into the
environment. TypeScript lives outside `web/engine/src/`, so the scoring
fingerprint and its change-note check are untouched, and outside every
`tsconfig.json` `include`, so `typecheck:all` is unaffected.

The write-up reports whatever the result is. "The prospective signal is
rho = 0.1 and here is the evidence that it is not more" is a real finding and
gets written up as one.

## Sequence of work

- [ ] 1. Feature extractor + label builder over all 27 comps; dataset stats.
  (~95 ms/track measured, so the whole corpus is ~10 min single-threaded.)
- [ ] 2. Weather fetch, run locally to clear the rate limit; ERA5 first,
  archived forecast behind it. Commit both trees to the archive.
- [ ] 3. Baselines. If baseline 4 says placement is nearly all pilot identity,
  that reframes the rest and gets said before any network is trained.
- [ ] 4. Retrospective MLP → the reference number.
- [ ] 5. Prospective MLP → the experiment.
- [ ] 6. Sequence branch, as a delta on both.
- [ ] 7. Ablations: track-only / +task / +ERA5 / +archived forecast; and the
  per-block feature-group ablation.
- [ ] 8. Write-up, commit, draft PR on each repo.

## Risks, stated up front

- **~5 500 samples is small** for anything deep. Hence: a small MLP first,
  heavy regularisation, comp-grouped CV, and a sequence branch only if the
  features earn it.
- **Two-thirds of the corpus is three comp series.** A fold holding out
  Forbes is holding out flatlands aerotow HG; ρ will vary a lot across folds
  and every number gets reported per fold, not just pooled.
- **The archived-forecast host is rate-limited from the cloud container**,
  whose proxy IP is shared. The fetch step therefore runs from a local
  machine, and the store it fills is committed to the archive so it is paid
  for once. The design still degrades to ERA5 rather than stalling, and the
  forecast arm is reported separately either way.
- **Retrospective leakage that isn't leakage.** Distance flown obviously
  predicts GAP points. That is not a bug in the experiment; it is why
  baseline 2 exists and why the prospective framing carries the weight.
